proxy/proxy.py

```diff
@@ -0,0 +1,121 @@
+        async with trio.open_nursery() as nursery:
+            while True:
+                try:
+                    apns_server = apns.APNSConnection(nursery)
+                    await apns_server._connect_socket()
+                    self.server = apns_server.sock
+
+                    nursery.start_soon(self.proxy, True)
+                    nursery.start_soon(self.proxy, False)
+
+                    break # Will only happen if there is no exception
+                except Exception:
+                    logging.error("Unable to start proxy, trying again...")
+                    await trio.sleep(1)
+
+    
+
+    async def proxy(self, to_server: bool):
+        if to_server:
+            from_stream = self.client
+            to_stream = self.server
+        else:
+            from_stream = self.server
+            to_stream = self.client
+        while True:
+            payload = await apns.APNSPayload.read_from_stream(from_stream)
+            payload = self.tamper(payload, to_server)
+            self.log(payload, to_server)
+            await payload.write_to_stream(to_stream)
+
+    def log(self, payload: apns.APNSPayload, to_server: bool):
+        printer.print_payload(payload, to_server)
+        # if to_server:
+        #     logging.info(f"-> {payload}")
+        # else:
+        #     logging.info(f"<- {payload}")
+        
+    def tamper(self, payload: apns.APNSPayload, to_server) -> apns.APNSPayload:
+        #if not to_server:
+        #    payload = self.tamper_lookup_keys(payload)
+
+        return payload
+
+    def tamper_lookup_keys(self, payload: apns.APNSPayload) -> apns.APNSPayload:
+        if payload.id == 0xA: # Notification
+            if payload.fields_with_id(2)[0].value == sha1(b"com.apple.madrid").digest(): # Topic
+                if payload.fields_with_id(3)[0].value is not None: # Body
+                    body = payload.fields_with_id(3)[0].value
+                    body = plistlib.loads(body)
+                    if body['c'] == 97: # Lookup response
+                        resp = gzip.decompress(body["b"]) # HTTP body
+                        resp = plistlib.loads(resp)
+
+                        # Replace public keys
+                        for r in resp["results"].keys():
+                            for i in range(len(resp["results"][r]["identities"])):
+                                if "client-data" in resp["results"][r]["identities"][i]:
+                                    resp["results"][r]["identities"][i]["client-data"]["public-message-identity-key"] = b"REDACTED"
+                        
+                        resp = gzip.compress(plistlib.dumps(resp, fmt=plistlib.FMT_BINARY), mtime=0)
+                        body["b"] = resp
+                    body = plistlib.dumps(body, fmt=plistlib.FMT_BINARY)
+                    for f in range(len(payload.fields)):
+                        if payload.fields[f].id == 3:
+                            payload.fields[f].value = body
```

Comment:
```suggestion
                    for field in payload.fields:
                        if field.id == 3:
                            field.value = body
```

Reply:
again, maybe I'm misunderstanding python, but I thought that was necessary so that it changed the original object and not a copy?